src/tests/JIT/Intrinsics/MemoryMarshalGetArrayDataReference.cs

```diff
@@ -0,0 +1,151 @@
+
+            IsFalse(Unsafe.IsNullRef(ref MemoryMarshal.GetArrayDataReference(Array.Empty<byte>())));
+            IsFalse(Unsafe.IsNullRef(ref MemoryMarshal.GetArrayDataReference(Array.Empty<string>())));
+            IsFalse(Unsafe.IsNullRef(ref MemoryMarshal.GetArrayDataReference(Array.Empty<Half>())));
+            IsFalse(Unsafe.IsNullRef(ref MemoryMarshal.GetArrayDataReference(Array.Empty<Vector128<byte>>())));
+            IsFalse(Unsafe.IsNullRef(ref MemoryMarshal.GetArrayDataReference(Array.Empty<StructWithByte>())));
+            IsFalse(Unsafe.IsNullRef(ref MemoryMarshal.GetArrayDataReference(Array.Empty<SimpleEnum>())));
+            IsFalse(Unsafe.IsNullRef(ref MemoryMarshal.GetArrayDataReference(Array.Empty<GenericStruct<byte>>())));
+            IsFalse(Unsafe.IsNullRef(ref MemoryMarshal.GetArrayDataReference(Array.Empty<GenericStruct<string>>())));
+
+            IsFalse(Unsafe.IsNullRef(ref ptrByte(Array.Empty<byte>())));
+            IsFalse(Unsafe.IsNullRef(ref ptrString(Array.Empty<string>())));
+
+            ThrowsNRE(() => { _ = ref MemoryMarshal.GetArrayDataReference<byte>(null); });
+            ThrowsNRE(() => { _ = ref MemoryMarshal.GetArrayDataReference<string>(null); });
+            ThrowsNRE(() => { _ = ref MemoryMarshal.GetArrayDataReference<Half>(null); });
+            ThrowsNRE(() => { _ = ref MemoryMarshal.GetArrayDataReference<Vector128<byte>>(null); });
+            ThrowsNRE(() => { _ = ref MemoryMarshal.GetArrayDataReference<StructWithByte>(null); });
+            ThrowsNRE(() => { _ = ref MemoryMarshal.GetArrayDataReference<SimpleEnum>(null); });
+            ThrowsNRE(() => { _ = ref MemoryMarshal.GetArrayDataReference<GenericStruct<byte>>(null); });
+            ThrowsNRE(() => { _ = ref MemoryMarshal.GetArrayDataReference<GenericStruct<string>>(null); });
+
+            ThrowsNRE(() => { _ = ref ptrByte(null); });
+            ThrowsNRE(() => { _ = ref ptrString(null); });
+
+            // from https://github.com/dotnet/runtime/issues/58312#issuecomment-993491291
+            [MethodImpl(MethodImplOptions.NoInlining)]
+            static int Problem1(StructWithByte[] a)
+            {
+                MemoryMarshal.GetArrayDataReference(a).Byte = 1;
+
+                a[0].Byte = 2;
+
+                return MemoryMarshal.GetArrayDataReference(a).Byte;
+            }
+
+            Equals(Problem1(new StructWithByte[] { new StructWithByte { Byte = 1 } }), 2);
+
+            [MethodImpl(MethodImplOptions.NoInlining)]
+            static int Problem2(byte[] a)
+            {
+                if (MemoryMarshal.GetArrayDataReference(a) == 1)
+                {
+                    a[0] = 2;
+                    if (MemoryMarshal.GetArrayDataReference(a) == 1)
+                    {
+                        return -1;
+                    }
+                }
+
+                return 0;
+            }
+
+            Equals(Problem2(new byte[] { 1 }), 0);
```

Comment:
I see the Problem1 and Problem2 tests failing with:
```
/__w/1/s/src/tests/JIT/Intrinsics/MemoryMarshalGetArrayDataReference.cs:L65 test failed (expected: equal, actual: 2-2).
/__w/1/s/src/tests/JIT/Intrinsics/MemoryMarshalGetArrayDataReference.cs:L82 test failed (expected: equal, actual: 0-0).
```
both before and after the change that suggest the problem is somewhere else.

Does the aliasing problem that was mentioned in the original comment actually repro in observable way on current main?

Reply:
The tests are missing an ! on Equals there, as you can see the actual values are the same here. I'm going to fix it tomorrow together with the missing NREs and invalid IndexOutOfRanges.

As far as the tests here, @SingleAccretion told me on discord that those should still happen on main, haven't verified it though.